docs/source/examples.daisy_chain_mlip_fitting.ipynb

```diff
@@ -0,0 +1,726 @@
+    "# Descriptor string, just as it would go into quip.\n",
+    "# dictionary can have a descriptor per species, e.g. \n",
+    "# descriptor = {\n",
+    "#   \"H\": \"soap ...\",\n",
+    "#   \"C\": \"soap ...\"}\n",
+    "# `None` for dictionary keys just means that the same descriptor is used \n",
+    "# for all elements. \n",
+    "descriptor =   {\n",
+    "        None: \"soap l_max=3 n_max=6 cutoff=4 delta=1 covariance_type=dot_product zeta=4 atom_gaussian_width=0.3\"\n",
+    "    }\n",
+    "\n",
+    "# this function isn't parallelised here, but can be\n",
+    "# by setting WFL_NUM_PYTHON_SUBPROCESSES or\n",
+    "# WFL_EXPYRE_INFO\n",
+    "md_soap_local = wfl.descriptors.quippy.calc(\n",
+    "    inputs=md_sample,\n",
+    "    outputs=outputs,\n",
+    "    descs=descriptor,\n",
+    "    key=descriptor_key,\n",
+    "    per_atom=True\n",
+    ")\n",
+    "\n",
+    "def get_average_soap(at, descriptor_key):\n",
+    "    at_desc = at.arrays.pop(descriptor_key)\n",
+    "    at_desc = np.sum(at_desc, axis=0)\n",
+    "    at_desc /= np.linalg.norm(at_desc)\n",
+    "    at.info[descriptor_key] = at_desc \n",
+    "    return at\n",
+    "\n",
+    "md_soap_global = wfl.map.run(\n",
+    "    inputs = md_soap_local,\n",
+    "    outputs = OutputSpec(), \n",
+    "    map_func = get_average_soap, \n",
+    "    args = [descriptor_key])"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Sub-select with CUR\n",
+    "\n",
+    "Select diverse structures for training and testing sets with CUR. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outputs = OutputSpec(\"4.ch.rdkit.md.traj.soap.cur_selection.xyz\")\n",
+    "cur_selection = wfl.select.by_descriptor.CUR_conf_global(\n",
+    "    inputs=md_soap_global,\n",
+    "    outputs=outputs,\n",
+    "    num=100,                    # target number of structures to pick\n",
+    "    at_descs_info_key=\"SOAP\")\n",
+    "    \n",
+    "\n",
+    "train_fname = \"5.1.train.xyz\"\n",
+    "test_fname = \"5.2.test.xyz\"\n",
+    "gap_fname='gap.xml'\n",
+    "\n",
+    "def process(at):\n",
+    "    at.cell = [50, 50, 50]\n",
+    "    del at.info[\"SOAP\"]\n",
+    "    return at\n",
+    "\n",
+    "\n",
+    "processed_cur_selection = wfl.map.run(\n",
+    "    cur_selection,\n",
+    "    OutputSpec(),\n",
+    "    map_func = process)\n",
+    "\n"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "label and save training and testing sets"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train_inputs = ConfigSet([list(processed_cur_selection)[0::2], read(isolated_at_fname, \":\")])\n",
+    "test_inputs = ConfigSet(list(processed_cur_selection)[1::2])\n",
```

Comment:
Do you think it's worth making (and how easy would it be) `ConfigSet` more list-like? In this case, make it suscriptable?

Reply:
I'll think about it.  It's coded as a generator because I really wanted to be able to use it for large datasets and not pull them all into memory at the same time, but that also prevents subscripting.

Reply:
Not as bad as I thought, if we're willing to accept that it could be inefficient, since it still has to go through all the configurations to the end (and this version, at least, doesn't support negative values like `-10:` for the last 10.

Can you want to try adding something this method to `ConfigSet`?
```
from itertools import islice
.
.
.
  def __getitem__(self, arg):
      if isinstance(arg, slice):
          return islice(self, arg.start, arg.stop, arg.step)
      elif isinstance(arg, int):
          return list(islice(self, arg, arg + 1, 1))[0]
      else:
          raise TypeError
```

Reply:
Had to modify the first return, but otherwise works!

Reply:
Is there a reason you converted the returned generators to lists? I don't generally like that, and the caller can always cast it into a list if they need an explicit list.

Reply:
It was throwing an error along the lines of "`islice` is not suscriptable". Withy my change it should be returning a `ConfigSet`. Please overwrite my changes if there's a better solution.

Reply:
If you want it to return *a ConfigSet* that's a different can of worms, which can also be done, but isn't what either of us implemented.  My version returns a single atomic config if the subscript is an int, or an iterable of configs is it's a slice. Yours returns a list if you pass a slice argument. Exactly what did you run that gave that error?  

If you tried to make it into a new `ConfigSet`, that wouldn't work, because `ConfigSet` needs a list, not an arbitrary iterable.  The current design is that you need to know that this constraint on creating `ConfigSet` exists, and that the subscript with a slice returns an iterable, not a list, and cast it yourself.  You can always iterate over the subscripts configset yourself and it'll work fine, or pass it to anything that can accept an iterable.

I can be convinced that the subscript with a slice should return a ConfigSet, or that you should be able to create a ConfigSet from an arbitrary iterable.  But returning a list strikes me as weird.  If we do decide that it should return a ConfigSet, we have to think about whether, if the original ConfigSet referred to files, whether the new one should refer to the same files but just skip some configs, or whether it should contain a distinct copy of the configs (presumably in memory).

Reply:
I think the (very slight) use-simplification I was looking for would mean making a ConfigSet from an arbitrary iterable, in addition to current `__getitem__`. But I don't think the added complexity is worth it after all (`list(iterable)` is perfectly fine). So let's leave it and I'll undo the changes from this thread?

Reply:
I agree with simpler.  

The ConfigSet from an arbitrary iterable isn't actually that complicated - the only problem is detecting what kind of input you passed `ConfigSet()`, but strings and `Atoms` are also iterable, so you can't just iterate and hope for the best. We could probably figure out sensible semantics if we wanted to.

Reply:
I think let's separate this from this PR. I have opened issue #229. Please resolve this if that sounds good.